rfcs/rfc-colin-states-mapping-to-lear.md

```diff
@@ -0,0 +1,127 @@
+}
+```
+
+An invalid business state transition map will be implemented as a python dictionary or database table.  This map can be used in conjunction with the existing [ALLOWABLE_FILINGS](https://github.com/bcgov/lear/blob/e850f9a22672910db6e4ceb2f1ddb9437541a86f/legal-api/src/legal_api/services/authz.py#L96) dictionary as well as any other conditions to determine what filings are allowed for a business at a given point in time.
+
+_Example of what invalid state transition might look like_
+``` python
+INVALID_STATE_TRANSITION_FILINGS: Final = {
+    'staff': {
+        Business.State.ACTIVE: { # current business state
+            'restoration': { # target filing
+                # invalid target filing if filing that transitioned business to its current state is in list
+                'fullRestoration': ['registration', 'incorporationApplication'], 
+                'limitedRestoration': ['registration', 'incorporationApplication'], 
+            }
+            ...
+        },
+        Business.State.HISTORICAL: {
+            'restoration': { 
+                'fullRestoration': ['almagmation', 'continuationOut', 'continuationIn'], 
+                'limitedRestoration': ['almagmation', 'continuationOut', 'continuationIn'],
+            },
+            'putBackOn': ['almagmation', 'withdrawal']
+            ...
+        },
+        Business.State.LIQUIDATION: {
+            'restoration': {
+                'fullRestoration': ['liquidation'], 
+                'limitedRestoration': ['liquidation'], 
+            },
+            ...
+        }
+    },
+    'client': {
+        Business.State.ACTIVE: { # current business state
+            ...
+        },
+        Business.State.HISTORICAL: {
+            ...
+        },
+        Business.State.LIQUIDATION: {
+            'restoration': {
+                'fullRestoration': ['liquidation'], 
+                'limitedRestoration': ['liquidation'], 
+            },
+            'putBackOn': ['liquidation']
+            ...
+        }
+    }        
+    
+}
+```
+
+_Allowable state transitions matrix in COLIN for restoration and put back on filings_
+![colin_allowable_filings.png](rfc-colin-states-mapping-to-lear/colin_allowable_filings.png)
+
+Filing validation logic will be updated to include logic to ensure that only valid business state transitions are allowed. 
+
+All filings resulting in business state transitions will need to populate the businesses.state_filing_id field.  This is used to determine if future business state transitions are permitted.
+
+Filings.meta_data field for any filings where a business state transitions occurs will include from/to state.
+
+API consumers such as the FE with a need to determine which filing transitioned the business to its current state will retrieve the business state filing provided by the business response of the GET business endpoint.  For example, this can be used to determine if an ACTIVE business is currently in _limited restoration_.
```

Comment:
I'm thinking API consumers such as the FE will need their own mapping table to determine what data to display / actions are allowed depending on the current state. This is business logic. It would ideal to keep any business logic out of the FE.

Any ideas? Should the prior "allowable filings" idea be expanded so the BE can inform the FE what to display/allow?

Reply:
That is what the `currentlyActionableFilingTypes` is intended for in the example business response at the start of the "Detailed Designed" section.  My thought is the FE can use this to determine what filings are possible given the current state of the business.

And yes, the prior "allowable filings" logic will be updated to integrate the invalid state transition piece.

Reply:
```
GET https://legal-api-dev.apps.silver.devops.gov.bc.ca/api/v2/businesses/CP0001199

{
    "business": {
        "legalName": "XYZ TEST CORP.",
        "legalType": "BC",
        "state": "HISTORICAL",
        "currentlyActionableFilingTypes": ["putBackOn", "restoration"],
        ...
}
```

Reply:
My plan is to always populate `currentlyActionableFilingTypes` with a complete list of filings the business can currently action.  We already have the `ALLOWABLE_FILINGS` map in the BE that contains that list.  Just need to run another check to filter out the invalid filings given the current state of the business using `INVALID_STATE_TRANSITION_FILINGS`.

Reply:
OK.

Here are the kinds of things the UI needs to know if they are allowed:
https://github.com/bcgov/business-filings-ui/blob/main/src/mixins/allowable-actions-mixin.ts

I think the currentlyActionableFilingTypes won't give us all that info.

Reply:
I know that talking about allowable actions seems out of scope for this RFC, but it would be good to not require extra business logic in the FE. In fact, it would be ideal to remove business logic. I think this RFC dove-tails nicely with that.

Reply:
I think we can support most of the file based actions in https://github.com/bcgov/business-filings-ui/blob/main/src/mixins/allowable-actions-mixin.ts without too much work.  We just need to add in `hasBlocker` and `hasBlockerExceptStaffApproval` logic.  Other conditions are supported by `ALLOWABLE_FILINGS` and `INVALID_STATE_TRANSITION_FILINGS`.  

I guess if we really wanted to we could also support the non-filing based actions eventually too.  Which makes me think I should re-structure business response to something like:

``` json
{
    "business": {
        "legalName": "XYZ TEST CORP.",
        "legalType": "BC",
        "state": "HISTORICAL",
        "currentlyAllowableActions": {
            "actions": ["addDetailComment", "addStaffComment", "viewChangeCompanyInfo"],
            "filingActions": {
                "restoration": ["fullRestoration", "limitedRestoration"], 
                "filingSubtypes": {
                   "restoration": ["fullRestoration", "limitedRestoration"]
                }   
            }
        }        
    }
}
```

Reply:
That would be \*awesome\* !

Reply:
Can we remove some of the stutter?
```
currentlyAllowableActions
    actions
    filingActions
```
```
allowableActions
    admin
    filing
```
and why is ```fullRestoration``` in there 2x?

Reply:
Updated as you suggested @thorwolpert.  `fullRestoration` in there 2x was accidental.

``` json
{
    "business": {
        "legalName": "XYZ TEST CORP.",
        "legalType": "BC",
        "state": "HISTORICAL",
        "allowableActions": {
            "admin": ["addDetailComment", "addStaffComment", "viewChangeCompanyInfo"]
            "filing": {
                "types": ["restoration", "putBackOn"],
                "subTypes": {
                   "restoration": ["fullRestoration", "limitedRestoration"]
                }
            }
        }
        ...
}
```

Reply:
As I started this conversation and my question has been answered, I'm resolving it. Feel free to open a new conversation if there are still/new unanswered questions.